pkg/controller/infrastructure/terraform_reconciler.go

```diff
@@ -234,3 +216,79 @@ func (r *TerraformReconciler) Delete(ctx context.Context, infra *extensionsv1alp
 func NoOpStateInitializer(_ context.Context, _ client.Client, _, _ string, _ *metav1.OwnerReference) error {
 	return nil
 }
+
+func (r *TerraformReconciler) getClientFactory(ctx context.Context, infra *extensionsv1alpha1.Infrastructure, cluster *controller.Cluster) (azureclient.Factory, error) {
+	cloudProfile, err := helper.CloudProfileConfigFromCluster(cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	var cloudConfiguration *azure.CloudConfiguration
+	if cloudProfile != nil {
+		cloudConfiguration = cloudProfile.CloudConfiguration
+	}
+
+	azCloudConfiguration, err := azureclient.AzureCloudConfigurationFromCloudConfiguration(cloudConfiguration)
+	if err != nil {
+		return nil, err
+	}
+
+	return DefaultAzureClientFactoryFunc(
+		ctx,
+		r.Client,
+		infra.Spec.SecretRef,
+		false,
+		azureclient.WithCloudConfiguration(azCloudConfiguration),
+	)
+}
+
+func (r *TerraformReconciler) cleanResourceGroupIfNeeded(ctx context.Context, infra *extensionsv1alpha1.Infrastructure, cluster *controller.Cluster, cfg *azure.InfrastructureConfig) error {
+	var err error
+	// skip operations on user resource groups
+	if cfg.ResourceGroup != nil {
+		return nil
+	}
+	// skip operations if we are not creating the resource group for the first time.
+	if lastOp := infra.Status.LastOperation; lastOp == nil || lastOp.Type != gardencorev1beta1.LastOperationTypeCreate {
+		return nil
+	}
+
+	status, err := helper.InfrastructureStatusFromInfrastructure(infra)
+	if err != nil {
+		return err
+	}
+
+	rgName := infrastructure.ShootResourceGroupName(infra, cfg, status)
+
+	clientFactory, err := r.getClientFactory(ctx, infra, cluster)
+	if err != nil {
+		return err
+	}
+
+	resourceGroupClient, err := clientFactory.Group()
+	if err != nil {
+		return err
+	}
+	if ok, err := resourceGroupClient.CheckExistence(ctx, rgName); err != nil {
+		return err
+	} else if !ok {
+		return nil
+	}
+
+	resourceClient, err := clientFactory.Resource()
+	if err != nil {
+		return err
+	}
+	res, err := resourceClient.ListByResourceGroup(ctx, rgName)
+	if client.IgnoreNotFound(err) != nil {
```

Comment:
Thats an interesting case - the resource group being absent after we explicitly checked for its existence. I agree that it's not worth to make a fuss about it at that point, but it might be worth a log-line if that case occurs.